RecoLocalCalo/HcalRecAlgos/src/MahiGPU.cu

```diff
@@ -0,0 +1,1901 @@
+        // update accum
+        #pragma unroll
+        for (int counter=iL; counter<NSAMPLES; counter++)
+            reg_b_tmp[counter] -= x_prev * reg_L[counter];
+
+        // load the next column of cholesky
+        #pragma unroll
+        for (int counter=iL; counter<NSAMPLES; counter++)
+            reg_L[counter] = matrixL(counter, iL);
+
+        // compute the next x for M(iL, icol)
+        x_prev = reg_b_tmp[iL] / reg_L[iL];
+
+        // store the result value
+        reg_b[iL] = x_prev;
+    }
+}
+
+// TODO: add active bxs
+template<typename MatrixType, typename VectorType>
+__device__
+void fnnls(
+        MatrixType const& AtA,
+        VectorType const& Atb,
+        VectorType& solution,
+        int& npassive,
+        ColumnVector<VectorType::RowsAtCompileTime, int> &pulseOffsets,
+        MapSymM<float, VectorType::RowsAtCompileTime> &matrixL,
+        double const eps,
+        int const maxIterations) {
+    // constants
+    constexpr auto NPULSES = VectorType::RowsAtCompileTime;
+
+    // to keep track of where to terminate if converged
+    Eigen::Index w_max_idx_prev = 0;
+    float w_max_prev = 0;
+    auto eps_to_use = eps;
+    bool recompute = false;
+
+    // used throughout
+    VectorType s;
+    float reg_b[NPULSES];
+    //float matrixLStorage[MapSymM<float, NPULSES>::total];
+    //MapSymM<float, NPULSES> matrixL{matrixLStorage};
+
+    int iter = 0;
+    while (true) {
+        if (iter > 0 || npassive==0) {
+            auto const nactive = NPULSES - npassive;
+            // exit if there are no more pulses to constrain
+            if (nactive==0) break;
+
+            // compute the gradient
+            //w.tail(nactive) = Atb.tail(nactive) - (AtA * solution).tail(nactive);
+            Eigen::Index w_max_idx;
+            float w_max = -std::numeric_limits<float>::max();
+            for (int icol=npassive; icol<NPULSES; icol++) {
+                auto const icol_real = pulseOffsets(icol);
+                auto const atb = Atb(icol_real);
+                float sum = 0;
+                #pragma unroll
+                for (int counter=0; counter<NPULSES; counter++)
+                    sum += counter > icol_real
+                        ? AtA(counter, icol_real) * solution(counter)
+                        : AtA(icol_real, counter) * solution(counter);
+
+                auto const w = atb - sum;
+                if (w > w_max) {
+                    w_max = w;
+                    w_max_idx = icol - npassive;
+                }
+            }
+
+            // check for convergence
+            if (w_max<eps_to_use || 
+                w_max_idx==w_max_idx_prev && w_max==w_max_prev)
+                break;
+
+            if (iter >= maxIterations) break;
+
+            w_max_prev = w_max;
+            w_max_idx_prev = w_max_idx;
+
+            // move index to the right part of the vector
+            w_max_idx += npassive;
+
+            Eigen::numext::swap(pulseOffsets.coeffRef(npassive),
```

Comment:
Q: how you avoided to swap all the vectors (and do only the fitted amplitude) ?
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc#L476-L484
If I do in the CPU version all the quantities are wrong.